src/main/java/de/tum/in/www1/artemis/domain/Exercise.java

```diff
@@ -597,6 +594,59 @@ public void filterForCourseDashboard(List<StudentParticipation> participations,
         }
     }
 
+    /**
+     * Filter for appropriate submission. Relevance in the following order:
+     * - submission with rated result
+     * - submission with unrated result (late submission)
+     * - no submission with any result > latest submission
+     *
+     * @param submissions that need to be filtered
+     * @return filtered submission
+     */
+    private Submission findAppropriateSubmission(Set<Submission> submissions) {
+        List<Submission> submissionsWithRatedResult = new ArrayList<>();
+        List<Submission> submissionsWithUnratedResult = new ArrayList<>();
+        List<Submission> submissionsWithoutResult = new ArrayList<>();
+
+        if (submissions != null) {
+            for (Submission submission : submissions) {
+                Result result = submission.getResult();
+                if (result != null) {
+                    if (result.isRated()) {
+                        submissionsWithRatedResult.add(submission);
+                    }
+                    else {
+                        submissionsWithUnratedResult.add(submission);
+                    }
+                }
+                else {
+                    submissionsWithoutResult.add(submission);
+                }
+            }
+        }
+
+        Comparator<Submission> comparator = ((Submission s1, Submission s2) -> {
+            if (s1.getSubmissionDate() == null || s2.getSubmissionDate() == null)
+                return 0;
+            return s1.getSubmissionDate().compareTo(s2.getSubmissionDate());
+        });
+
+        if (submissionsWithRatedResult.size() > 0) {
```

Comment:
Comparator, sorting and getting the result can be moved to a separate method.  From here you will only need to pass the list of submissions and check the result. This will remove the code duplication introduced here.

Reply:
Good point. Thank you!